rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+preprocessing_stage.adapt(data_sample)
+
+model = keras.Sequential([
+    preprocessing_stage,
+    keras.layers.Dense(10, activation='softmax'),
+])
+model.fit(data, targets, epochs=10)
+```
+
+
+### Base class: `PreprocessingLayer`
+
+All preprocessing layers inherit from a base class: `PreprocessingLayer`, which itself inherits from `Layer`.
+
+This class presents a few key differences compared to regular layers:
+
+**Separate training mechanism**
+
+The internal state of a `PreprocessingLayer` is not affected by backpropagation: all of its weights are non-trainable. A `PreprocessingLayer` has to be trained in a separate step, as follow:
+
+```python
+preprocessing_layer.adapt(data_sample)
+```
+
+**Possible non-differentiability**
+
+Processing layers extend Keras by allowing preprocessing to be part of the model. Unlike existing layers, these computations are not always differentiable, e.g. both `Discretize` and `VectorizeText` are non-differentiable.
+
+As a result, all preprocessing layers are treated as frozen when used as part of a model. In addition, if a non-differentiable layer is used in the middle of a model (rather than at the start), the model will raise an exception related to differentiability when trying to compute gradients (e.g. as part of `fit`).
+
+
+### New layers
+
+- `PreprocessingLayer` base class: implements shared logic, in particular the `adapt` method for setting the state of the layer.
+- `PreprocessingStage` class: makes it possible to chain multiple preprocessing layers together while training them in one single `adapt` call (by doing cascading training of the underlying layers).
+- `Normalization`: normalizes data feature-wise by subtracting the mean of some sample dataset and dividing by the variance.
+- `Discretization`: transforms continuous data into one-hot encoded binary vectors representing the different "bins" that the continuous data belongs to.
+- `TextVectorization`: transforms string data into either dense vectors (e.g. TF-IDF transform) or sequences of token indices (e.g. to be passed to an `Embedding` layer).
+
+
+## Design details
+
+### Detailed layer signatures
+
+#### PreprocessingLayer
+
+```python
+def adapt(self, data, reset_state=True):
+    """Fits the state of the preprocessing layer to the data being passed.
+
+    Arguments:
+        data: The data to train on. It can be passed either as a tf.data Dataset,
+            or as a numpy array (or a dict or list of arrays in case of multi-input
+            preprocessing stages).
+        reset_state: Optional argument specifying whether to clear the state of the
+            layer at the start of the call to `adapt`, or whether to start from
+            the existing state. This argument may not be relevant to all
+            preprocessing layers: a subclass of PreprocessingLayer may chose to
+            only implement `adapt(self, data)`.
+    """
+```
+
+#### PrepocessingStage
```

Comment:
Two more most commonly used preprocessing layers are:
1. Resize
2. ToTensor => Converting an image tensor (0 to 255) into a tensor (0 to 1) values.

Should we support it in this proposal?

Reply:
Also, these 3 pre-processing operations (ToTensor, Resize, Normalize) are very well suited to be handed off to the accelerator for significant performance boost. 

I had done a similar work in MXNet Gluon - https://medium.com/apache-mxnet/fusing-data-transformations-with-neural-network-for-faster-inference-and-simpler-production-2be50457be8a

Reply:
Your `ToTensor` seems like it would be covered by the normalization layer (it's just a divide by `255` for each channel).

The issue with a layer that does resizing is that it could not process a batch of tensors. All layers (including preprocessing layers) are meant for batch-wise processing: one batch of N samples in, one batch of N samples out.

The problem here being that if you have images of different sizes, you couldn't pack them into a batch tensor.  So by the time images enter a layer, they should all have the same size. So resizing should happen in the tf.data pipeline.

It could work to resize a batch of images that have already been standardized to a common size, but that is not very useful.

Reply:
1. ToTensor => Agreed, can be put in Normalization layer. But, it should be an option for users? If not, here we are assuming it will be always a an image with values 0..255.
2. Yes this is a big user problem, during inference specially, that they need to manage resizing. How about, resize preprocessing layer that can take list of tensors and resize and batchify the output into a batched tensor. By solving this, I think we will make the model fully self-sufficient to just take an image and give back results.